models/activities/action.go

```diff
@@ -450,17 +450,46 @@ func GetFeeds(ctx context.Context, opts GetFeedsOptions) (ActionList, int64, err
 		return nil, 0, err
 	}
 
-	sess := db.GetEngine(ctx).Where(cond).
-		Select("`action`.*"). // this line will avoid select other joined table's columns
-		Join("INNER", "repository", "`repository`.id = `action`.repo_id")
+	actions := make([]*Action, 0, opts.PageSize)
+	var count int64
 
-	opts.SetDefaultValues()
-	sess = db.SetSessionPagination(sess, &opts)
+	if opts.Page < 10 { // TODO: why it's 10 but other values? It's an experience value.
+		sess := db.GetEngine(ctx).Where(cond).
+			Select("`action`.*"). // this line will avoid select other joined table's columns
+			Join("INNER", "repository", "`repository`.id = `action`.repo_id")
 
-	actions := make([]*Action, 0, opts.PageSize)
-	count, err := sess.Desc("`action`.created_unix").FindAndCount(&actions)
-	if err != nil {
-		return nil, 0, fmt.Errorf("FindAndCount: %w", err)
+		opts.SetDefaultValues()
+		sess = db.SetSessionPagination(sess, &opts)
+
+		count, err = sess.Desc("`action`.created_unix").FindAndCount(&actions)
+		if err != nil {
+			return nil, 0, fmt.Errorf("FindAndCount: %w", err)
+		}
+	} else {
+		// First, only query which IDs are necessary, and only then query all actions to speed up the overall query
+		sess := db.GetEngine(ctx).Where(cond).
+			Select("`action`.id").
+			Join("INNER", "repository", "`repository`.id = `action`.repo_id")
+
+		opts.SetDefaultValues()
+		sess = db.SetSessionPagination(sess, &opts)
+
+		actionIDs := make([]int64, 0, opts.PageSize)
+		if err := sess.Table("action").Desc("`action`.created_unix").Find(&actionIDs); err != nil {
+			return nil, 0, fmt.Errorf("Find: %w", err)
```

Comment:
The error info is too simple ( same to L491 ). Not good for debug.

Reply:
49cbb2fc3ff63877e4924d94f0cf26c6285c0985